### ISSUE TITLE
sphinx.ext.viewcode could not work as expected on parallel build

### ISSUE THREAD

**Issue by 0xMashiro** [sphinx-doc/sphinx#8933]
**Describe the bug**
I'm using sphinx to build the document for the project [MegEngine](https://github.com/MegEngine/MegEngine/tree/master/imperative/python/megengine). 

 In [Docs](https://github.com/MegEngine/Tutorials/tree/docs-transfer/source) repo, most APIs are organized manually for some reasons:

```rst
.. autosummary::
   :toctree: api
   :nosignatures:

   cos
   sin
   tan
   acos
   asin
   atan
   atan2
   cosh
   sinh
   tanh
   acosh
   asinh
   atanh
```
As you can see, several functions are implemented here:

<img width="1514" alt="WeChat3bebe346a97db2dad70a63161e7f0ff9" src="https://user-images.githubusercontent.com/21091736/109069228-e7baa500-772b-11eb-97a2-8e3f868b07c2.png">

Everything works for `cosh()`:

<img width="1586" alt="WeChata2050cf6586bb572655ecd991b785331" src="https://user-images.githubusercontent.com/21091736/109069492-4122d400-772c-11eb-9ef0-adda152e0fe8.png">

<img width="2014" alt="WeChate65d067246afe5f0d378e5776c672f1d" src="https://user-images.githubusercontent.com/21091736/109069582-60216600-772c-11eb-8ddf-f82843035f62.png">

![WeChat91eaedabab3eb86bbf83a80dc7574b44](https://user-images.githubusercontent.com/21091736/109070124-11c09700-772d-11eb-8ea6-2068a260a651.png)

But it doesn't work for some other APIs like `asin()` `acos()` ...

Seems `viewcode-back` class is not generated for every API.

**To Reproduce**

**I'm sure that `sphinx.ext.viewcode`  not works as expected.**
You can do it with these steps to reproduce the behavior with a GitHub workflow file: https://github.com/MegEngine/Tutorials/blob/main/.github/workflows/sphinx-build.yml

**Expected behavior**
copied Class/Function source codes HTML pages can be highlighted and add links to documents correctly.

**Your project**
[MegEngine](https://github.com/MegEngine/MegEngine) / [Docs](https://github.com/MegEngine/Tutorials/tree/docs-transfer/source)

**Environment info**
- OS: Ubuntu 18.04
- Python version: 3.6/3.7
- Sphinx version: sphinx-build 3.5.1
- Sphinx extensions:  sphinx.ext.viewcode
- Extra tools: -

**Comment by MegChai** [comment]
Even though I tried to use the full path instead of using the alias, there are still many missing highlights and jumps. My previous information may be inaccurate, and I cannot determine whether it is the alias caused or other reasons.

**Comment by MegChai** [comment]
As you can see, several functions are implemented here:

<img width="1514" alt="WeChat3bebe346a97db2dad70a63161e7f0ff9" src="https://user-images.githubusercontent.com/21091736/109069228-e7baa500-772b-11eb-97a2-8e3f868b07c2.png">

Everything works for `cosh()`:

<img width="1586" alt="WeChata2050cf6586bb572655ecd991b785331" src="https://user-images.githubusercontent.com/21091736/109069492-4122d400-772c-11eb-9ef0-adda152e0fe8.png">

<img width="2014" alt="WeChate65d067246afe5f0d378e5776c672f1d" src="https://user-images.githubusercontent.com/21091736/109069582-60216600-772c-11eb-8ddf-f82843035f62.png">

But it doesn't work for eg: `asin()`. Just jumps to `../_modules/megengine/functional/elemwise.html#asin` URL 

![WeChat91eaedabab3eb86bbf83a80dc7574b44](https://user-images.githubusercontent.com/21091736/109070124-11c09700-772d-11eb-8ea6-2068a260a651.png)

Seems `viewcode-back` class is not generated for every API.

**Comment by tk0miya** [comment]
Now I checked your public document on https://megengine.org.cn/api/latest/_modules/megengine/functional/elemwise.html#cosh, but it seems working fine. Does this trouble happen in your local environment?

**Comment by 0xMashiro** [comment]
@tk0miya Thanks for your reply. Now our doc [website](https://megengine.org.cn/) is an old version(all APIs are generated by `sphinx.ext.autodoc` `automodule`) and it will be replaced soon. If you want to reproduce the problem locally, here is the source code: https://github.com/MegEngine/Tutorials/tree/docs-transfer
 
And I can provide a [Tutorial](https://github.com/MegEngine/Tutorials/blob/docs-transfer/source/developmet/contribute-to-docs.rst#%E5%85%8B%E9%9A%86%E6%96%87%E6%A1%A3%E6%BA%90%E7%A0%81%E5%88%B0%E6%9C%AC%E5%9C%B0) to build and deploy the document locally. But it was written in Chinese so I think you can use Google translation or look at this [file](https://github.com/MegEngine/Tutorials/blob/main/.github/workflows/sphinx-build.yml): 

```yaml
jobs:
  Build:
    name: Build html
    runs-on: ubuntu-latest
    steps:
      - name: Checkout document repp
        uses: actions/checkout@v2

      - name: Checkout theme repo
        uses: actions/checkout@v2
        with: 
          repository: MegEngine/pydata-sphinx-theme
          ref: dev
          fetch-depth: 1
          path: theme
    
      - name: Setup Python
        uses: actions/setup-python@v2
        with:
          python-version: 3.6

      - name: Cache Python packages
        id: cache-python
        uses: actions/cache@v2
        with:
          path: ${{ env.pythonLocation }}
          key: ${{ env.pythonLocation }}$-${{ hashFiles('requirements.txt') }}

      - name: Install MegEngine, Sphinx and other Python dependencies
        if: steps.cache-python.outputs.cache-hit != 'true'
        run: |
          python3 -m pip install --upgrade pip
          if [ -f requirements.txt ]; then pip install -r requirements.txt; fi
          pip install megengine
      - name: Install Pydata Sphinx theme
        run: python3 -m pip install ./theme

      - name: Install pandoc and other dependencies
        run: sudo apt-get install -y pandoc

      - name: Generate HTML files
        run: make html -W --keep-going
```

It's in the testing status for MegEngine 1.3 (not release now) so there will be lots of API missing warning if you build it with MegEngine v1.2.

**Comment by tk0miya** [comment]
I found the reason. It's a bug of the viewcode extension on parallel build mode. #8935 will resolve the error.

**Comment by 0xMashiro** [comment]
So efficient. 👍

**Comment by astrojuanlu** [comment]
@MegChai Hi! I see that you closed MegEngine/Documentation#49 but, from what I understand, MegEngine has Sphinx 3.5.4 pinned. We are investigating whether there is still a bug with `sphinx.ext.viewcode` accidentally introduced by #8935 (see #9210). Could you please try to build MegEngine docs with Sphinx 4.0 and report back the results? Thanks in advance!

**Comment by 0xMashiro** [comment]
@astrojuanlu Cool. `sphinx.ext.viewcode` works as expected now. Thanks for your work.

I fix the Sphinx version [because of other unadapted extensions](https://github.com/MegEngine/Documentation/commit/33aaf430848be409ab46e19733be40a3bfc6abb8)  and have a [workaround](https://github.com/mgaitan/sphinxcontrib-mermaid/issues/72) later. And I will start using Sphinx 4.0 in the next version of MegEngine documentation (see [WIP PR](https://github.com/MegEngine/Documentation/pull/163/commits/52326b98e2b643fe421fdc9276c4b8857d40fdef)).
